File: packages/renderer/src/compositor/payloads.ts
export type CompositorImageFormat = 'Png' | 'Jpeg';

export type VideoMetadata = {
	fps: number;
	width: number;
	height: number;
	durationInSeconds: number | null;
	codec: 'h264' | 'h265' | 'vp8' | 'vp9' | 'av1' | 'prores' | 'unknown';
	canPlayInVideoTag: boolean;
	supportsSeeking: boolean;
	colorSpace:
		| 'rgb'
		| 'bt601'
		| 'bt709'
		| 'bt2020-ncl'
		| 'bt2020-cl'
		| 'fcc'
		| 'bt470bg'
		| 'smpte170m'
		| 'smpte240m'
		| 'ycgco'
		| 'smpte2085'
		| 'chroma-derived-ncl'
		| 'chroma-derived-cl'
		| 'ictcp'
		| 'unknown';
	audioCodec:
		| null
		| 'opus'
		| 'aac'
		| 'mp3'
		| 'pcm-f16le'
		| 'pcm-f24le'
		| 'pcm-f32be'
		| 'pcm-s16be'
		| 'pcm-s16le'
		| 'pcm-f32le'
		| 'pcm-s32be'
		| 'pcm-s32le'
		| 'pcm-s64be'
		| 'pcm-s64le'
		| 'pcm-u16be'
		| 'pcm-u16le'
		| 'pcm-u24be'
		| 'pcm-u24le'
		| 'pcm-u32be'
		| 'pcm-u32le'
		| 'pcm-u8'
		| 'pcm-f64be'
		| 'pcm-s24be'
		| 'pcm-s24le'
		| 'pcm-s8'
		| 'pcm-s16be-planar'
		| 'pcm-s8-planar'
		| 'pcm-s24le-planar'
		| 'pcm-s32le-planar'
		| 'unknown';
	audioFileExtension: string | null;
	pixelFormat:
		| null
		| 'yuv420p'
		| 'yuyv422'
		| 'rgb24'
		| 'bgr24'
		| 'yuv422p'
		| 'yuv444p'
		| 'yuv410p'
		| 'yuv411p'
		| 'yuvj420p'
		| 'yuvj422p'
		| 'yuvj444p'
		| 'argb'
		| 'rgba'
		| 'abgr'
		| 'bgra'
		| 'yuv440p'
		| 'yuvj440p'
		| 'yuva420p'
		| 'yuv420p16le'
		| 'yuv420p16be'
		| 'yuv422p16le'
		| 'yuv422p16be'
		| 'yuv444p16le'
		| 'yuv444p16be'
		| 'yuv420p9be'
		| 'yuv420p9le'
		| 'yuv420p10be'
		| 'yuv420p10le'
		| 'yuv422p10be'
		| 'yuv422p10le'
		| 'yuv444p9be'
		| 'yuv444p9le'
		| 'yuv444p10be'
		| 'yuv444p10le'
		| 'yuv422p9be'
		| 'yuv422p9le'
		| 'yuva420p9be'
		| 'yuva420p9le'
		| 'yuva422p9be'
		| 'yuva422p9le'
		| 'yuva444p9be'
		| 'yuva444p9le'
		| 'yuva420p10be'
		| 'yuva420p10le'
		| 'yuva422p10be'
		| 'yuva422p10le'
		| 'yuva444p10be'
		| 'yuva444p10le'
		| 'yuva420p16be'
		| 'yuva420p16le'
		| 'yuva422p16be'
		| 'yuva422p16le'
		| 'yuva444p16be'
		| 'yuva444p16le'
		| 'yuva444p'
		| 'yuva422p'
		| 'yuv420p12be'
		| 'yuv420p12le'
		| 'yuv420p14be'
		| 'yuv420p14le'
		| 'yuv422p12be'
		| 'yuv422p12le'
		| 'yuv422p14be'
		| 'yuv422p14le'
		| 'yuv444p12be'
		| 'yuv444p12le'
		| 'yuv444p14be'
		| 'yuv444p14le'
		| 'yuvj411p'
		| 'yuv440p10le'
		| 'yuv440p10be'
		| 'yuv440p12le'
		| 'yuv440p12be'
		| 'yuv420p9'
		| 'yuv422p9'
		| 'yuv444p9'
		| 'yuv420p10'
		| 'yuv422p10'
		| 'yuv440p10'
		| 'yuv444p10'
		| 'yuv420p12'
		| 'yuv422p12'
		| 'yuv440p12'
		| 'yuv444p12'
		| 'yuv420p14'
		| 'yuv422p14'
		| 'yuv444p14'
		| 'yuv420p16'
		| 'yuv422p16'
		| 'yuv444p16'
		| 'yuva420p9'
		| 'yuva422p9'
		| 'yuva444p9'
		| 'yuva420p10'
		| 'yuva422p10'
		| 'yuva444p10'
		| 'yuva420p16'
		| 'yuva422p16'
		| 'yuva444p16'
		| 'yuva422p12be'
		| 'yuva422p12le'
		| 'yuva444p12be'
		| 'yuva444p12le'
		| 'unknown';
};

export type SilentPart = {
	startInSeconds: number;
	endInSeconds: number;
};

export type SilentParts = SilentPart[];

export type GetSilentPartsResponseRust = {
	silentParts: SilentParts;
	durationInSeconds: number;
};

export type GetSilentPartsResponse = GetSilentPartsResponseRust & {
	audibleParts: SilentParts;
};

export type CompositorCommand = {
	ExtractFrame: {
		src: string;
		original_src: string;
		time: number;
		transparent: boolean;
		tone_mapped: boolean;
	};
	GetSilences: {
		src: string;
		noiseThresholdInDecibels: number;
		minDurationInSeconds: number;
	};
	Echo: {
		message: string;
	};
	StartLongRunningProcess: {
		concurrency: number;
		maximum_frame_cache_size_in_bytes: number | null;
		verbose: boolean;
	};
	GetOpenVideoStats: {};
	DeliberatePanic: {};
	CloseAllVideos: {};
	FreeUpMemory: {
		remaining_bytes: number;
	};
	GetVideoMetadata: {src: string};
	ExtractAudio: {input_path: string; output_path: string};
	VideoMetadata: VideoMetadata;
};

export type CompositorCommandSerialized<T extends keyof CompositorCommand> = {
	nonce: string;
	payload: {
		type: T;
		params: CompositorCommand[T];
	};
};

export type ErrorPayload = {
	error: string;
	backtrace: string;
};

File: packages/compositor/rust/payloads.rs
// Must keep this file synced with payload.ts!

extern crate serde;
extern crate serde_json;

pub mod payloads {
    use crate::errors::ErrorWithBacktrace;
    use serde::{Deserialize, Serialize};

    #[derive(Serialize, Deserialize, Debug)]
    pub struct ImageLayer {
        pub src: String,
        pub x: u32,
        pub y: u32,
        pub width: u32,
        pub height: u32,
    }

    #[derive(Serialize, Deserialize, Debug)]
    pub struct SolidLayer {
        pub fill: [u8; 4],
        pub x: u32,
        pub y: u32,
        pub width: u32,
        pub height: u32,
    }

    #[derive(Serialize, Debug)]
    pub struct ErrorPayload {
        pub error: String,
        pub backtrace: String,
    }

    #[derive(Serialize, Deserialize, Debug)]
    pub enum ImageFormat {
        Png,
        Jpeg,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct ExtractFrameCommand {
        pub src: String,
        pub original_src: String,
        pub time: f64,
        pub transparent: bool,
        pub tone_mapped: bool,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    #[allow(non_snake_case)]
    pub struct GetSilences {
        pub src: String,
        pub noiseThresholdInDecibels: f64,
        pub minDurationInSeconds: f64,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct StartPayLoad {
        pub concurrency: usize,
        pub maximum_frame_cache_size_in_bytes: Option<u64>,
        pub verbose: bool,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct GetOpenVideoStats {}

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct DeliberatePanic {}

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct CloseAllVideos {}

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct OpenVideoStats {
        pub open_streams: usize,
        pub frames_in_cache: usize,
    }

    #[derive(Serialize, Deserialize, Debug)]
    pub enum KnownCodecs {
        #[serde(rename = "h264")]
        H264,
        #[serde(rename = "h265")]
        H265,
        #[serde(rename = "vp8")]
        Vp8,
        #[serde(rename = "vp9")]
        Vp9,
        #[serde(rename = "av1")]
        Av1,
        #[serde(rename = "prores")]
        ProRes,
        #[serde(rename = "unknown")]
        Unknown,
    }

    #[derive(Serialize, Deserialize, Debug)]
    pub enum KnownAudioCodecs {
        #[serde(rename = "opus")]
        Opus,
        #[serde(rename = "aac")]
        Aac,
        #[serde(rename = "mp3")]
        Mp3,
        #[serde(rename = "pcm-f16le")]
        PcmF16Le,
        #[serde(rename = "pcm-f24le")]
        PcmF24Le,
        #[serde(rename = "pcm-f32be")]
        PcmF32be,
        #[serde(rename = "pcm-s16be")]
        PcmS16be,
        #[serde(rename = "pcm-s16le")]
        PcmS16le,
        #[serde(rename = "pcm-f32le")]
        PcmF32le,
        #[serde(rename = "pcm-s32be")]
        PcmS32be,
        #[serde(rename = "pcm-s32le")]
        PcmS32le,
        #[serde(rename = "pcm-s64be")]
        PcmS64be,
        #[serde(rename = "pcm-s64le")]
        PcmS64le,
        #[serde(rename = "pcm-u16be")]
        PcmU16be,
        #[serde(rename = "pcm-u16le")]
        PcmU16le,
        #[serde(rename = "pcm-u24be")]
        PcmU24be,
        #[serde(rename = "pcm-u24le")]
        PcmU24le,
        #[serde(rename = "pcm-u32be")]
        PcmU32be,
        #[serde(rename = "pcm-u32le")]
        PcmU32le,
        #[serde(rename = "pcm-u8")]
        PcmU8,
        #[serde(rename = "pcm-f64be")]
        PcmF64be,
        #[serde(rename = "pcm-s24be")]
        PcmS24be,
        #[serde(rename = "pcm-s24le")]
        PcmS24le,
        #[serde(rename = "pcm-s8")]
        PcmS8,
        #[serde(rename = "pcm-s16be-planar")]
        PcmS16bePlanar,
        #[serde(rename = "pcm-s8-planar")]
        PcmS8Planar,
        #[serde(rename = "pcm-s24le-planar")]
        PcmS24lePlanar,
        #[serde(rename = "pcm-s32le-planar")]
        PcmS32lePlanar,
        #[serde(rename = "vorbis")]
        Vorbis,
        #[serde(rename = "unknown")]
        Unknown,
    }

    #[derive(Serialize, Deserialize, Debug)]
    pub enum KnownColorSpaces {
        #[serde(rename = "rgb")]
        RGB,
        #[serde(rename = "bt601")]
        BT601,
        #[serde(rename = "bt709")]
        BT709,
        #[serde(rename = "bt2020-ncl")]
        BT2020NCL,
        #[serde(rename = "bt2020-cl")]
        BT2020CL,
        #[serde(rename = "fcc")]
        FCC,
        #[serde(rename = "bt470bg")]
        BT470BG,
        #[serde(rename = "smpte170m")]
        SMPTE170M,
        #[serde(rename = "smpte240m")]
        SMPTE240M,
        #[serde(rename = "ycgco")]
        YCGCO,
        #[serde(rename = "smpte2085")]
        SMPTE2085,
        #[serde(rename = "chroma-derived-ncl")]
        CHROMADERIVEDNCL,
        #[serde(rename = "chroma-derived-cl")]
        CHROMADERIVEDCL,
        #[serde(rename = "ictcp")]
        ICTCP,
        #[serde(rename = "unknown")]
        Unknown,
    }

    #[derive(Serialize, Deserialize, Debug)]
    #[allow(non_snake_case)]
    pub struct VideoMetadata {
        pub fps: f32,
        pub width: u32,
        pub height: u32,
        pub durationInSeconds: Option<f64>,
        pub codec: KnownCodecs,
        pub canPlayInVideoTag: bool,
        pub supportsSeeking: bool,
        pub colorSpace: KnownColorSpaces,
        pub audioCodec: Option<KnownAudioCodecs>,
        pub audioFileExtension: Option<String>,
        pub pixelFormat: Option<String>,
    }

    #[derive(Serialize, Deserialize, Debug)]
    #[allow(non_snake_case)]
    pub struct SilentParts {
        pub startInSeconds: f64,
        pub endInSeconds: f64,
    }

    #[derive(Serialize, Deserialize, Debug)]
    #[allow(non_snake_case)]
    pub struct GetSilentPartsResponse {
        pub silentParts: Vec<SilentParts>,
        pub durationInSeconds: f64,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct FreeUpMemory {
        pub remaining_bytes: u64,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct EchoPayload {
        pub message: String,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct GetVideoMetadata {
        pub src: String,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct ExtractAudio {
        pub input_path: String,
        pub output_path: String,
    }

    #[derive(Clone, Serialize, Deserialize, Debug)]
    pub struct DeleteFramesFromCache {
        pub maximum_frame_cache_size_in_bytes: u64,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct Eof {}

    #[derive(Serialize, Deserialize, Debug, Clone)]
    #[serde(tag = "type", content = "params")]
    pub enum CliInputCommandPayload {
        ExtractFrame(ExtractFrameCommand),
        StartLongRunningProcess(StartPayLoad),
        DeliberatePanic(DeliberatePanic),
        CloseAllVideos(CloseAllVideos),
        GetOpenVideoStats(GetOpenVideoStats),
        FreeUpMemory(FreeUpMemory),
        Echo(EchoPayload),
        GetVideoMetadata(GetVideoMetadata),
        GetSilences(GetSilences),
        ExtractAudio(ExtractAudio),
        Eof(Eof),
        DeleteFramesFromCache(DeleteFramesFromCache),
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct CliInputCommand {
        pub payload: CliInputCommandPayload,
        pub nonce: String,
    }

    #[derive(Serialize, Deserialize, Debug, Clone)]
    pub struct CliInputAndMaxCacheSize {
        pub cli_input: CliInputCommand,
        pub max_cache_size: u64,
    }

    pub fn parse_cli(json: &str) -> Result<CliInputCommand, ErrorWithBacktrace> {
        let cli_input: CliInputCommand = serde_json::from_str(json)?;

        return Ok(cli_input);
    }
}
